pkg/arvo/app/chat-hook.hoon

```diff
@@ -51,8 +54,137 @@
     ==
   ++  on-save   !>(state)
   ++  on-load
-    |=  old=vase
-    `this(state !<(state-zero old))
+    |=  =old=vase
+    =/  old  !<(versioned-state old-vase)
+    ?:  ?=(%1 -.old)
+      [~ this(state old)]
+    ::  path structure ugprade logic
+    ::
+    :_  this(state [%1 +.old])
+    %-  zing
+    ^-  (list (list card))
+    %+  turn
+      %~  tap  in
+      %^  scry:cc  (set path)
+        %chat-store
+      /keys
+    |^  |=  chat=path
+        ^-  (list card)
+        =/  host=ship  (slav %p (snag 0 chat))
+        =/  newp=permission  (unify-permissions chat)
+        =/  old-group=path  [%chat chat]
+        =/  new-group=path  [%'~' chat]
+        ;:  weld
+          :~  (delete-group host (snoc old-group %read))
+              (delete-group host (snoc old-group %write))
+          ==
+        ::
+          (create-group new-group who.newp)
+          (hookup-group new-group kind.newp)
+          [(record-group new-group chat)]~
+        ==
+    ::
+    ++  unify-permissions
+      |=  chat=path
+      ^-  permission
+      =/  read=(unit permission)   (get-permission chat %read)
+      =/  write=(unit permission)  (get-permission chat %write)
+      ?.  &(?=(^ read) ?=(^ write))
+        ~&  [%missing-permission chat read=?=(~ read) write=?=(~ write)]
+        [%white [(slav %p (snag 0 chat)) ~ ~]]
+      ::  village: exclusive to writers
+      ::
+      ?:  &(?=(%white kind.u.read) ?=(%white kind.u.write))
```

Comment:
97 through 109 would be much clearer if it was a ?+ branching off of [kind.u.read kind.u.write]

Reply:
Good call, this is much better. (Kinda sad the type system isn't smart enough to let you write this as a `?-`, but w/e.)